### ISSUE TITLE
java.lang.RuntimeException: Block with no successors is not a tail! raised during building CompleteBlockGraph

### ISSUE THREAD

**Issue by jikk** [Sable/soot#345]
Hi, I'm trying to build CompleteBlockGraph, but I encountered the following error.
Would you please take a look at the following issue? 

Thanks a lot for your help in advance.

* Command to perform analysis

```
java -Xss50m -Xmx1500m -classpath /PATHNAME/soot_analysis/lib/soot-trunk.jar:/PATHNAME/soot_analysis/lib/AXMLPrinter2.jar:/PATHNAME/soot_analysis/lib/baksmali-2.0.3.jar: soot.tools.CFGViewer --graph=CompleteBlockGraph -d out/ -android-jars /PATHNAME/soot_analysis/platforms/ -allow-phantom-refs -src-prec apk -ire -f J -process-dir /PATHNAME/smali/apk_repo/com.opentable.apk
```

* Error message and stack trace

```
Generate dot file in out//void a(com.google.android.gms.maps.model.Tile,android.os.Parcel,int).dot
Transforming com.opentable.activities.search.SearchResults$SearchResultObserver...
Generate dot file in out//void <init>(com.opentable.activities.search.SearchResults).dot
Generate dot file in out//void <init>(com.opentable.activities.search.SearchResults,com.opentable.activities.search.SearchResults$1).dot
Generate dot file in out//void handleSearchError(com.android.volley.VolleyError).dot
Generate dot file in out//void handleSearchSuccess().dot
java.lang.RuntimeException: Block with no successors is not a tail!: Block 34:
[preds: ] [succs: ]
nop;

  at soot.toolkits.graph.BlockGraph.buildBlocks(BlockGraph.java:285)
  at soot.toolkits.graph.ExceptionalBlockGraph.buildBlocks(ExceptionalBlockGraph.java:119)
  at soot.toolkits.graph.BlockGraph.<init>(BlockGraph.java:72)
  at soot.toolkits.graph.ExceptionalBlockGraph.<init>(ExceptionalBlockGraph.java:99)
  at soot.toolkits.graph.CompleteBlockGraph.<init>(CompleteBlockGraph.java:52)
  at soot.util.cfgcmd.CFGGraphType$8.buildGraph(CFGGraphType.java:195)
  at soot.tools.CFGViewer.print_cfg(CFGViewer.java:246)
  at soot.tools.CFGViewer.internalTransform(CFGViewer.java:76)
  at soot.BodyTransformer.transform(BodyTransformer.java:51)
  at soot.Transform.apply(Transform.java:105)
  at soot.BodyPack.internalApply(BodyPack.java:49)
  at soot.Pack.apply(Pack.java:126)
  at soot.PackManager.runBodyPacks(PackManager.java:911)
  at soot.PackManager.runBodyPacks(PackManager.java:606)
  at soot.PackManager.runBodyPacks(PackManager.java:507)
  at soot.PackManager.runPacksNormally(PackManager.java:484)
  at soot.PackManager.runPacks(PackManager.java:391)
  at soot.Main.run(Main.java:244)
  at soot.Main.main(Main.java:152)
  at soot.tools.CFGViewer.main(CFGViewer.java:101)
  ```

**Comment by ericbodden** [comment]
Hi Jee.

Would it be possible for you to have Soot print out the Jimple code for the method in question? It looks like the code may be incorrectly structured.

About that: Are you modifying the Jimple in any way by extending Soot?

Cheers,
Eric

**Comment by jikk** [comment]
I think so. It worked fine with the following other configurations.

* BriefBlockGraph

```
 java -Xss50m -Xmx1500m -classpath /PATHNAME/soot_analysis/lib/soot-trunk.jar:/PATHNAME/soot_analysis/lib/AXMLPrinter2.jar:/PATHNAME/soot_analysis/lib/baksmali-2.0.3.jar: soot.tools.CFGViewer --graph=BriefBlockGraph -d out/ -android-jars /PATHNAME/soot_analysis/platforms/ -allow-phantom-refs -src-prec apk -ire -f J -process-dir /PATHNAME/smali/apk_repo/com.opentable.apk
```

* soot.Main

```
java -Xss50m -Xmx1500m -classpath /PATHNAME/soot_analysis/lib/soot-trunk.jar:/PATHNAME/soot_analysis/lib/AXMLPrinter2.jar:/PATHNAME/soot_analysis/lib/baksmali-2.0.3.jar: soot.Main -d out/ -android-jars /PATHNAME/soot_analysis/platforms/ -allow-phantom-refs -src-prec apk -ire -f J -process-dir /PATHNAME/smali/apk_repo/com.opentable.apk
```

Above commands worked fine and produced expected outputs. Is there anything that I can get more for you?

Thanks for the prompt response.

**Comment by ericbodden** [comment]
Hi again.

The problem is that we don't have your com.opentable.apk here to reproduce this.

Could you run Soot with the "-f J" option and attach to this bug the Jimple code for the method

void handleSearchSuccess()

in class

com.opentable.activities.search.SearchResults$SearchResultObserver

That seems to be the method which causes the error.

Cheers,
Eric

**Comment by StevenArzt** [comment]
Fixed with 97fbbd5c0e6fcadba17c8ce46c8fba0e5b7b5cf7
